src/main.ts

```diff
@@ -232,83 +230,234 @@ class AutoEntities extends LitElement {
     }
     entities = entities.filter(Boolean);
 
-    const include_filters = await Promise.all(
-      (this._config.filter?.include ?? []).map(async (filter) => {
-        if (filter.type) return async () => [filter as LovelaceRowConfig];
-
-        const filters = await get_filter(this.hass, filter);
-        const sorter = filter.sort?.method
-          ? await get_sorter(this.hass, filter.sort)
-          : (x) => x;
-
-        const post_process = async (entity) =>
-          await process_entity(
-            this.hass,
-            { ...entity, ...filter.options },
-            entity.entity
-          );
-
-        return async (entities: EntityList) => {
-          let add = entities.filter(filters);
-          // Filter-local sort
-          add = await sorter(add);
-          // Filter-local pagination
-          if (filter.sort?.count || filter.sort?.first) {
-            const start = filter.sort?.first ?? 0;
-            const count = filter.sort?.count ?? Infinity;
-            add = add.slice(start, start + count);
+    if (this._config.filter?.include) {
+      const all_entities = Object.keys(this.hass.states).map(format);
+      for (const filter of this._config.filter.include) {
+        if (filter.type) {
+          entities.push(filter);
+          continue;
+        }
+
+        let add: EntityList = [];
+        for (const entity of all_entities) {
+          if (await filter_entity(this.hass, filter, entity.entity))
+            add.push(
+              JSON.parse(
+                JSON.stringify({ ...entity, ...filter.options }).replace(
+                  /this.entity_id/g,
+                  entity.entity
+                )
+              )
+            );
+        }
+
+        if (filter.sort) {
+          await getEntities(this.hass);
+          await getDevices(this.hass);
+          await getAreas(this.hass);
+          add = add.sort(get_sorter(this.hass, filter.sort));
+          if (filter.sort.count ?? filter.sort.first) {
+            const start = filter.sort.first ?? 0;
+            add = add.slice(start, start + (filter.sort.count ?? Infinity));
           }
-          add = await Promise.all(add.map(post_process));
-          return add;
+        }
+        entities = entities.concat(add);
+      }
+    }
+
+    // TODO: Add tests for exclusions
+    if (this._config.filter?.exclude) {
+      for (const filter of this._config.filter.exclude) {
+        const newEntities = [];
+        for (const entity of entities) {
+          if (
+            entity.entity === undefined ||
+            !(await filter_entity(this.hass, filter, entity.entity))
+          )
+            newEntities.push(entity);
+        }
+        entities = newEntities;
+      }
+    }
+
+    if (this._config.names) {
+      const { transforms } = this._config.names;
```

Comment:
As code is lengthly it should be placed in its own file for ease of readability of main code and maintenance of names and transforms. Follow how sort is done.